### PR TITLE
feat: Cosmos SDK `v0.46.0`

### PR DESCRIPTION
**Pull request by aljo242**
Updates the Cosmos SDK to `v0.46.0`.  Please review this PR after the other onces I have submitted are completed.

This PR builds off of those, and will be much more manageable to review if the other PRs are already merged.  

- makes changes according to the [changelog](https://github.com/cosmos/cosmos-sdk/blob/release/v0.46.x/CHANGELOG.md)

This PR introduces temporary `replace` clauses in our `go.mod` for the following packages:
- `github.com/cosmos/ibc-go/v5 => github.com/aljo242/ibc-go/v5 v5.1.0`
- `github.com/tendermint/fundraising => github.com/aljo242/fundraising v0.1.2-0.20220726203752-c39abf16b02d`

Once the respective repos are updated, we can remove these lines.  I have an open [PR](https://github.com/tendermint/fundraising/pull/192) for `x/fundraising` and `ibc-go` is near completion.

Part of #909 

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Include tests
- [ ] Respect code style and lint
- [ ] Update documentation (*.md) (if needed)

### CODE REVIEW

**Comment by lubtd** [comment]
What difference with https://github.com/tendermint/spn/pull/906?

**Comment by aljo242** [comment]
> What difference with #906?
> 


That PR included all of the changes that I made in the previous PRs to prepare for `v0.46.0`.  It would have been a mess to review. 

This PR also fixes tests that were broken by the upgrade.

> Is it with the final imports for IBC and fundraising?

These are the final imports for a temporary solution.  My update to `fundraising` works, but I'm just waiting for them to approve.  We will still need to make a PR later that removes the `replace` lines in the `go.mod` once the dependencies are fixed.

File: x/mint/keeper/integration_test.go
Comment (lubtd) [inline x/mint/keeper/integration_test.go:98]: Is it a requirement for this upgrade or part of https://github.com/tendermint/spn/issues/852?

Looks like this utility could be moved to `testutil`

**Review by lubtd — COMMENTED** [review]
Great work

I put some initial comments. I need to do some experimentation and check further the IBC part

File: x/mint/keeper/integration_test.go
Comment (aljo242) [reply x/mint/keeper/integration_test.go:98]: This is required for the upgrade.  The test will now fail without a validator in the state.  I agree though - moving to `testutil`.

**Comment by lubtd** [comment]
I'm expected `ignite serve` to fail currently but not `ignite g proto-go`

Seems like there are some changes for proto
```
cosmos/base/v1beta1/coin.proto:32:8: Option "(cosmos_proto.scalar)" unknown. Ensure that your proto definition file imports the proto which defines the option.
cosmos/base/v1beta1/coin.proto:37:19: Option "(cosmos_proto.scalar)" unknown. Ensure that your proto definition file imports the proto which defines the option.
cosmos/base/v1beta1/coin.proto:42:19: Option "(cosmos_proto.scalar)" unknown. Ensure that your proto definition file imports the proto which defines the option.
campaign/special_allocations.proto:7:1: Import "cosmos/base/v1beta1/coin.proto" was not found or had errors.
campaign/special_allocations.proto:10:12: "cosmos.base.v1beta1.Coin" is not defined.
campaign/special_allocations.proto:15:12: "cosmos.base.v1beta1.Coin" is not defined.
campaign/campaign.proto:7:1: Import "cosmos/base/v1beta1/coin.proto" was not found or had errors.
campaign/campaign.proto:8:1: Import "campaign/special_allocations.proto" was not found or had errors.
campaign/campaign.proto:17:12: "cosmos.base.v1beta1.Coin" is not defined.
campaign/campaign.proto:22:12: "cosmos.base.v1beta1.Coin" is not defined.
campaign/campaign.proto:27:3: "SpecialAllocations" is not defined.
```

**Comment by aljo242** [comment]
> I'm expected `ignite serve` to fail currently but not `ignite g proto-go`
> 
> Seems like there are some changes for proto
> 
> ```
> cosmos/base/v1beta1/coin.proto:32:8: Option "(cosmos_proto.scalar)" unknown. Ensure that your proto definition file imports the proto which defines the option.
> cosmos/base/v1beta1/coin.proto:37:19: Option "(cosmos_proto.scalar)" unknown. Ensure that your proto definition file imports the proto which defines the option.
> cosmos/base/v1beta1/coin.proto:42:19: Option "(cosmos_proto.scalar)" unknown. Ensure that your proto definition file imports the proto which defines the option.
> campaign/special_allocations.proto:7:1: Import "cosmos/base/v1beta1/coin.proto" was not found or had errors.
> campaign/special_allocations.proto:10:12: "cosmos.base.v1beta1.Coin" is not defined.
> campaign/special_allocations.proto:15:12: "cosmos.base.v1beta1.Coin" is not defined.
> campaign/campaign.proto:7:1: Import "cosmos/base/v1beta1/coin.proto" was not found or had errors.
> campaign/campaign.proto:8:1: Import "campaign/special_allocations.proto" was not found or had errors.
> campaign/campaign.proto:17:12: "cosmos.base.v1beta1.Coin" is not defined.
> campaign/campaign.proto:22:12: "cosmos.base.v1beta1.Coin" is not defined.
> campaign/campaign.proto:27:3: "SpecialAllocations" is not defined.
> ```

It looks like we needed to add a dependency for `cosmos-proto` using this command:

```shell
go get github.com/regen-network/cosmos-proto@latest          
```

Let me know if this works for you now.

**Comment by aljo242** [comment]
There is some dependency on the `cosmos-proto` package now.  I think that the `ignite` command may need to be modified to accommodate that

**Comment by aljo242** [comment]
If you run

```shell
go get github.com/regen-network/cosmos-proto@latest              
ignite chain serve --proto-all-modules --clear-cache --force-reset -v
```

proto compilation works for me.  Does this work for you as well?

**Comment by aljo242** [comment]
https://github.com/cosmos/cosmos-sdk/pull/11452/files this PR might be helpful

**Comment by lubtd** [comment]
> If you run
> 
> ```shell
> go get github.com/regen-network/cosmos-proto@latest              
> ignite chain serve --proto-all-modules --clear-cache --force-reset -v
> ```
> 
> proto compilation works for me. Does this work for you as well?

Nope, I have the same error

**Comment by lubtd** [comment]
Supporting the proto changes will be part of https://github.com/ignite/cli/issues/2156

**Comment by aljo242** [comment]
> Supporting the proto changes will be part of [ignite/cli#2156](https://github.com/ignite/cli/issues/2156)

Sounds good.  If that's the only outstanding issue, are we good to try to get this approved?

**Comment by lubtd** [comment]
> > Supporting the proto changes will be part of [ignite/cli#2156](https://github.com/ignite/cli/issues/2156)
> 
> Sounds good. If that's the only outstanding issue, are we good to try to get this approved?

I think this is better if we can merge it once we can run it with `ignite c serve` and use `ignite c build`.

If we need the new version for `spn` dependency in `ignite`, maybe we can temporarily update `spn` dep in `ignite` to this branch?

**Comment by aljo242** [comment]
> > > Supporting the proto changes will be part of [ignite/cli#2156](https://github.com/ignite/cli/issues/2156)
> > 
> > 
> > Sounds good. If that's the only outstanding issue, are we good to try to get this approved?
> 
> I think this is better if we can merge it once we can run it with `ignite c serve` and use `ignite c build`.
> 
> If we need the new version for `spn` dependency in `ignite`, maybe we can temporarily update `spn` dep in `ignite` to this branch?

I think it would be better the other way around.  If this upgrade fully works, then the only remaining issue is that `ignite` is out of date.  So it makes more sense to upgrade this, then fix `ignite` after.

After all, we have now modified our dependencies so that:
- `ignite` imports `spn`
- `spn` no longer imports `ignite`

**Comment by aljo242** [comment]
> What does it change if we do `go get github.com/tendermint/spn@6bff6689d24b74fd3d6c2786db475e5250708ef9 ` on Ignite?
> 
> This will remove the circular dep, then we just need to update again `spn` dep on Ignite once [ignite/cli#2673](https://github.com/ignite/cli/issues/2673) is finished. I can't test everything on the changes without using Ignite

Good point - I was thinking this would introduce a fork in the `go.mod` but my branch is on here.  I'll go ahead and do that.

**Comment by ilgooz** [comment]
There could be one problem with that because that commit hash wouldn't be guaranteed to be kept forever on Github.

The reason for this, once we merge this PR, all the commits here will be squashed into a single commit and the commit that has mentioned will not have a reference on develop PR and also in the base branch of this PR since it'll be deleted after merging.

I suggest merging this PR first and using the commit hash added to the develop branch instead.

**Comment by ilgooz** [comment]
But if we plan to change the commit hash after we finish work then np.

**Comment by aljo242** [comment]
Ah - I see what you mean.  I just need some approvals and we can merge

**Comment by lubtd** [comment]
> There could be one problem with that because that commit hash wouldn't be guaranteed to be kept forever on Github.
> 
> The reason for this, once we merge this PR, all the commits here will be squashed into a single commit and the commit that has mentioned will not have a reference on develop PR and also in the base branch of this PR since the branch will be deleted after merging.
> 
> I suggest merging this PR first and using the commit hash added to the develop branch instead.

It will be temporary, the time the `serve` command works CLI side for `0.46` we don't need to wait for the full `0.46` PR to be merged CLI side

**Comment by lubtd** [comment]
And as soon as this PR is merged we can change it

**Comment by ilgooz** [comment]
In that case there should be no problem. 👍

File: scripts/protocgen2.sh
Comment (ilgooz) [inline scripts/protocgen2.sh:1]: not a fan of these scripts actually. 

 Ignite CLI automatically gets the proto files and uses the same versions in your go mod, scripts always use latest versions which is not ideal

Comment (lubtd) [reply scripts/protocgen2.sh:1]: Version of `cosmos-proto`?

Comment (tbruyelle) [reply scripts/protocgen2.sh:1]: The official way to track tool dependencies in a module (not sure this applies well in the execution context of this script) :
https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module

Also mentioned here : https://github.com/ignite/cli/pull/2680#discussion_r933353893

Comment (ilgooz) [reply scripts/protocgen2.sh:1]: > Version of cosmos-proto?


Not sure if I understand what do you mean.

Users shouldn't need to maintain a script file to build code from proto. It's not the most pleasant way of doin it and also makes it hard to keep track of versions. CLI on the other hand checks your go.mod and retrieves the correct versions for the proto files that your chain depends on for code generation, instead of relying on "latest"

Comment (lubtd) [reply scripts/protocgen2.sh:1]: We are not imposing anything, `spn` users can use the already built proto

For Ignite in general, I think we should let the freedom to the user to choose the tool they want to use, either Ignite by default or `buf` like many projects in the ecosystem do like `cosmos-sdk`

**Review by lubtd — APPROVED** [review]
Tested all the features

image

File: scripts/protocgen2.sh
Comment (lubtd) [reply scripts/protocgen2.sh:1]: Let's move forward?

We need the proto formatting tool in any case, we can revert for proto generation if there is issues with this solution in the future.

We can also open a specific issue if there is a strong opinion about only relying on Ignite for proto

**Pull request merged**
